core/src/message/init.rs

```diff
@@ -49,7 +49,7 @@ impl InitMessage {
     pub fn from_packet(id: MessageId, packet: InitPacket) -> Self {
         Self {
             id,
-            destination: packet.program_id,
+            destination: packet.destination(id),
```

Comment:
That's wrong. It definetely creates untraceable id duplicates. Program can use same salt and code id, as was used to create itself. The message id will remain same if the call described previously happens in `init`.

The idea was that from pallet-level you use a global program ids nonce, but from `Externalities` level you use `u32`/`u64` nonce.

Reply:
nonces should be local to the program that creates other programs

There should not be global id, as it will lead to parallelization problem (since everybody will be able to access global counter)

Reply:
> nonces should be local to the program that creates other programs
> 
> There should not be global nonce, as it will lead to parallelization problem (since everybody will be able to access global counter)

yeah, optimal solution is to follow message id generation except in-block counter for creating from user